### PR TITLE
udiskslinuxfilesystem: Use udev ID_FS_SIZE if available

### PR DESCRIPTION
**Pull request by tbzatek**
Latest addition to blkid as released in util-linux-2.39 exposes ID_FS_FSLASTBLOCK and ID_FS_FSSIZE keys carrying calculated filesystem size.

The legacy path of property value retrieval on demand is still in place in case the udev property is unavailable (or zero if unsupported filesystem type in libblkid).

### CODE REVIEW

**Comment by tbzatek** [comment]
Okay, looks like in XFS case the `ID_FS_SIZE` is the data part only while libblockdev reports also log data size calculated together. Using `ID_FS_LASTBLOCK * ID_FS_BLOCKSIZE` will give the latter expected number.

**Comment by tbzatek** [comment]
@mvollmer FYI just in case you see a failure in Cockpit tests on rawhide and c9s, hope I haven't broken anything.

**Review by vojtechtrefny — APPROVED** [review]

**Comment by mvollmer** [comment]
> Okay, looks like in XFS case the `ID_FS_SIZE` is the data part only while libblockdev reports also log data size calculated together. Using `ID_FS_LASTBLOCK * ID_FS_BLOCKSIZE` will give the latter expected number.

For Cockpit, we want to know whether a filesystem is using its block device completely, so that we can warn the user about it.  `ID_FS_LASTBLOCK * ID_FS_BLOCKSIZE` should work, but `ID_FS_SIZE` will very likely trigger the warning spuriously, and any attempt fixing it will look like it has no effect.

**Comment by mvollmer** [comment]
> > Okay, looks like in XFS case the `ID_FS_SIZE` is the data part only while libblockdev reports also log data size calculated together. Using `ID_FS_LASTBLOCK * ID_FS_BLOCKSIZE` will give the latter expected number.
> 
> For Cockpit, we want to know whether a filesystem is using its block device completely, so that we can warn the user about it. `ID_FS_LASTBLOCK * ID_FS_BLOCKSIZE` should work, but `ID_FS_SIZE` will very likely trigger the warning spuriously, and any attempt fixing it will look like it has no effect.

This is also documented as part of the UDisks D-Bus API:

```
Size: The size of the filesystem.  This is the amount of
      bytes used on the block device.  If this is smaller than
      org.freedesktop.Udisks2.Block.Size, then the filesystem can
      be made larger with Resize.
```

So using `ID_FS_SIZE` for xfs would be a bug, no?

**Comment by tbzatek** [comment]
> ```
> Size: The size of the filesystem.  This is the amount of
>       bytes used on the block device.  If this is smaller than
>       org.freedesktop.Udisks2.Block.Size, then the filesystem can
>       be made larger with Resize.
> ```
> 
> So using `ID_FS_SIZE` for xfs would be a bug, no?

You're right, let's ditch it. Reading through util-linux source, it seems the calculation is indeed different by purpose.
See #1115

File: src/udiskslinuxfilesystem.c [hunk 1]
```diff
@@ -331,8 +334,18 @@ udisks_linux_filesystem_update (UDisksLinuxFilesystem  *filesystem,
...
+   * reporting.
+   */
+  filesystem->cached_fs_size = g_udev_device_get_property_as_uint64 (device->udev_device, "ID_FS_LASTBLOCK") *
+                               g_udev_device_get_property_as_uint64 (device->udev_device, "ID_FS_BLOCKSIZE");
+  if (filesystem->cached_fs_size == 0)
+    filesystem->cached_fs_size = g_udev_device_get_property_as_uint64 (device->udev_device, "ID_FS_SIZE");
+  /* The Size property is hacked to be retrieved on-demand, only need to
+   * notify subscribers that it has changed.
    */
   invalidate_property (filesystem, "Size");
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/udiskslinuxfilesystem.c:350]: That was my previous attempt too, however once the GObject class property gets overriden in `udisks_linux_filesystem_class_init()`, the generated interface skeleton doesn't see any value and the original GDBus machinery doesn't work.

What we can do is to put another `PropertiesChanged` signal emittance in place with the actual value, not sure if it's worth it though. Any d-bus client should be prepared to handle both cases no matter if a property value was transmitted or not. Still, does the current approach causes any trouble to your Cockpit dbus-to-web translations? The difference is that clients need to make another method call to get the value (which GDBusProxy would do itself anyway, to fill its property cache).

**Pull request merged**